src/electron/routing_service.ts

```diff
@@ -223,22 +225,60 @@ export class RoutingDaemon {
 
 //#region routing service installation
 
+/**
+ * Execute arbitary shell `command` as root.
+ * @param command Any valid shell command(s).
+ */
 function executeCommandAsRoot(command: string): Promise<void> {
   return new Promise<void>((resolve, reject) => {
-    sudo.exec(command, {name: 'Outline'}, sudoError => {
+    sudo.exec(command, {name: 'Outline'}, (sudoError, stdout, stderr) => {
+      console.info(stdout);
+      console.error(stderr);
+      // This error message is an un-exported constant defined here:
+      //   - https://github.com/jorangreef/sudo-prompt/blob/v9.2.1/index.js#L670
+      const PERMISSION_DENIED = 'did not grant permission';
       if (sudoError) {
-        // NOTE: The script could have terminated with an error - see the comment in
-        //       sudo-prompt's typings definition.
-        console.error('failed to execute command as root: ', sudoError);
-        reject(toErrorCode(new NoAdminPermissions(sudoError.message)));
+        if (sudoError.message?.includes(PERMISSION_DENIED)) {
+          console.error('user rejected to run command as root: ', sudoError);
+          reject(ErrorCode.NO_ADMIN_PERMISSIONS);
+        } else {
+          console.error('command is running as root but failed: ', sudoError);
+          reject(ErrorCode.UNEXPECTED);
+        }
       } else {
         resolve();
       }
     });
   });
 }
 
-function installWindowsRoutingServices(): Promise<void> {
+/**
+ * Execute arbitary shell `command` as a new child process.
+ * @param command Any valid shell command(s).
+ * @returns The standard output of the `command`.
+ */
+async function executeShellCommand(command: string): Promise<string> {
+  const exec = promisify(child_process.exec);
+  const {stdout, stderr} = await exec(command);
+  console.info(stdout);
+  console.error(stderr);
+  return stdout;
+}
+
+/**
+ * By default, root user is not able to connect to X Server, but it is required
+ * if we need to run Outline-Client (even in CLI mode) under root. This function let
+ * the caller specify whether root user is allowed to access X server. It will also
+ * return whether root is allowed before (so we can revert the change we made).
+ */
+async function configureRootXServerAccess(allowRootAccess: boolean): Promise<boolean> {
+  const ROOT_USER_ID = 'si:localuser:root';
+  const prevOut = (await executeShellCommand('xhost')) ?? '';
```

Comment:
Separate this into two functions: one to check the xhost status, and one to set it. To reason to interleave them.